### ISSUE TITLE
[Question] How to check if a VO is uninitialized?

### ISSUE THREAD

**Issue by CheloXL** [SteveDunn/Vogen#348]
### Describe the feature

For validation purposes, I need to check if some VO has a value or not. Right now, Vogen doesn't let me use == default as that always returns false.

Accessing the .Value property throws an exception, and I'm not using exceptions for validation, so that's not an option.

Since VOs comes from different places where I don't have control (asp actions, EF default values) I would really need a way to check if a VO is initialized or not without throwing an exception.

Maybe expose the field _isInitialized?

**Comment by Blackclaws** [comment]
I think what you're requesting is somewhat contrary to how ValueObjects are _supposed_ to work. When you have an instance of a ValueObject you're supposed to know its initialized. 

Now given your examples of sources where you don't have control. I'd recommend to take control somewhat back, for example by using nullable annotations. EFCore default values for Nullable fields are simply null, so you can check against that. 

Is there a specific reason you cannot be sure that your Values are valid and nullables wouldn't be a solution in your case?

Typically exceptions make sense in the case that a ValueObject isn't initialized or fails Validation defined by the Validation method. Is there a reason you're avoiding exceptions in this case?

**Comment by Blackclaws** [comment]
> I'm using nullables for that, but I didn't want to wrap the VO just for that.

Its not really wrapping it that much. Or are you using ValueType ValueObjects? I tend to use ReferenceType ValueObjects for exactly that reason that I don't have to do Value.Value when accessing it through a nullable variable. I find that there isn't really a benefit to using ValueType ValueObjects as instances are immutable anyway.

When using ReferenceTypes it should be much simpler to define your query parameters etc. as nullable and then just check whether they were set on the query to begin with.

If you want you can paste a minimal example and I'll see whether I can figure something out that wouldn't require you to either catch exception or access internal fields.

**Comment by jeffward01** [comment]
> Now given your examples of sources where you don't have control. I'd recommend to take control somewhat back, for example by using nullable annotations. EFCore default values for Nullable fields are simply null, so you can check against that.

From experience this is a little painful with EfCore.  Such as, usually, if I create a new **Child** object that is attached to a **Parent** object, I would only need to populate the data property, not the FK property.  If I do not populate the FK property, I get an error, so now I have to refactor my code to populate the FK property, or at least initialize it.

```csharp
public class Student
{
    public StudentId StuduentId { get; set; }
     
    // This is the FK of 'FavoriteTeacher'
    public FavoriteTeacherId FavoriteTeacherId { get; set; }

    // Usually i only populate this, but and EfCore is 'smart' enough to auto-populate the FK
    // I have found that if i do this, it will throw an error
    public virtual FavoriteTeacher FavoriteTeacher { get; set; }
}
```

Has this been your experience as well?

@CheloXL - What I have done in my code is wrap my code in a `try {...} catch {...}` statement.  Not ideal, i get it. But @Blackclaws is correct in what he says about the values of ValueObjects and null.  However, you are also correct with in some cases this can't be as 'controlled', so compromises have to be made.  My compromise was the try catch.

This code snippet is for an EfCore Entity. If the Id (Primary Key) is null or default, then the entity is 'new' (Transient):

```csharp

public abstract class Entity
{
    public virtual bool IsTransient()
    {
        try
        {
            if (this.Id == null)
            {
                return true;
            }

            bool isTransient = Equals(this.Id, default(TEntityKey));

            if (typeof(TEntityKey) != typeof(string))
            {
                return isTransient;
            }

            if (isTransient.IsFalse())
            {
                isTransient = string.IsNullOrWhiteSpace(this.Id?.ToString());
            }

            return isTransient;
        }
        catch (ValueObjectValidationException)
        {
            return true;
        }
        catch (Exception e)
        {
            Console.WriteLine(e);

            throw;
        }
    }
}

```

Basically, just wrap your code in the try catch, look for `ValueObjectValidationException`, and assume that if that is thrown then it is 'not initialized'.

Not a perfect solution, but it works and also keeps the principles of ValueObjects as @Blackclaws  described

**Comment by Blackclaws** [comment]
> From experience this is a little painful with EfCore. Such as, usually, if I create a new **Child** object that is attached to a **Parent** object, I would only need to populate the data property, not the FK property. If I do not populate the FK property, I get an error, so now I have to refactor my code to populate the FK property, or at least initialize it.

@jeffward01 Nullable, once again fixes your problem here rather easily. Let me give you a short example. I've made an example program for you here: https://github.com/Blackclaws/EfCoreVogenExample

The gist of it is, if you have entities with nullable properties:
```c#
[ValueObject<int>(Conversions.EfCoreValueConverter)]
public partial class TeacherId { }

[ValueObject<int>(Conversions.EfCoreValueConverter)]
public partial class StudentId { }

public class Teacher
{
    public TeacherId? Id { get; set; }
    public string TeacherName { get; init; } = "";
}

public class Student
{
    public StudentId? Id { get; set; }
    public Teacher? FavoriteTeacher { get; set; }
}

public class StudentWithFk
{
    public StudentId? Id { get; set; }
    public TeacherId? FavoriteTeacherId { get; set;  }
    public Teacher? FavoriteTeacher { get; set; }
}
```

And you set them up like this:
```c#
protected override void OnModelCreating(ModelBuilder modelBuilder)
    {
        /*modelBuilder.Entity<TestEntity>().HasMany<TestOwned>(x => x.OwnedInternal).WithOne(x => x.Parent)
            .HasForeignKey("ParentId");
        modelBuilder.Entity<TestOwned>().HasKey(new string[]
        {
            nameof(TestOwned.Id), "ParentId"
        });*/

        modelBuilder.Entity<Teacher>(teacherBuilder =>
        {
            teacherBuilder.HasKey(x => x.Id);
            teacherBuilder.Property(x => x.Id).HasConversion<TeacherId.EfCoreValueConverter>().ValueGeneratedOnAdd();
        });

        modelBuilder.Entity<Student>(studentBuilder =>
        {
            studentBuilder.HasKey(x => x.Id);
            studentBuilder.Property(x => x.Id).HasConversion<StudentId.EfCoreValueConverter>().ValueGeneratedOnAdd();
            studentBuilder.HasOne(x => x.FavoriteTeacher);
        });
        
        modelBuilder.Entity<StudentWithFk>(studentBuilder =>
        {
            studentBuilder.HasKey(x => x.Id);
            studentBuilder.Property(x => x.Id).HasConversion<StudentId.EfCoreValueConverter>().ValueGeneratedOnAdd();
            studentBuilder.HasOne(x => x.FavoriteTeacher).WithMany().HasForeignKey(x => x.FavoriteTeacherId);
        });
    }
```

Then this will do exactly what you asked for. And you can just check whether either the navigation property is null, the foreign key is null, or the ids of the entities themselves are null.

As a general rule, unless you have a special use case where you directly need to set the ForeignKeyId, I'd omit it from my C# classes to begin with.

**Comment by CheloXL** [comment]
@Blackclaws I think your are missing the point here. Yes, I can use a reference type and call it a day. But that doesn't address the main issue that is that I can't know if my VO is initialized or not without throwing an exception.

It's like if I have a reference type with null value and not having a way to check against null because you always get `false` (for reasons), and the only way to check if the reference is null is to wrap an accesor to some property in a try/catch and catch a NullReferenceException.

It works? Yes. It is the best/optimun way? I don't think so. The same applies to any struct in c#. Imagine to have to wrap any access to a DateTime in a try/catch just to check if it has the default value or not.

**Comment by Blackclaws** [comment]
> It works? Yes. It is the best/optimun way? I don't think so. The same applies to any struct in c#. Imagine to have to wrap any access to a DateTime in a try/catch just to check if it has the default value or not.

@CheloXL That's why nullables exist though. I get that the underlying issue is that, yes given a ValueObject of dubious origin (for example from reflection), you are not able to tell whether the object has been initialized or not by default. I still stand by my opinion that you should get rid of sources of dubious origin altogether. If you switch to nullable instances of ValueObjects you _should_ not get into the situation where you can't just do a check ``` if(valueObject is null) HandleError(); ``` and be sure afterwards that valueObject is initialized.

Now on to providing a solution that would actually solve your problem:

```
[ValueObject<string>()]
public partial class ValueObjectClass
{
    public bool Initialized => _isInitialized;
}
```

Just define the Property yourself, you have access to all private fields.

**Comment by CheloXL** [comment]
> 
> Now on to providing a solution that would actually solve your problem:
> 
> ```
> [ValueObject<string>()]
> public partial class ValueObjectClass
> {
>     public bool Initialized => _isInitialized;
> }
> ```
> 
> Just define the Property yourself, you have access to all private fields.

Ah... I always forget that I actually have access to the generated code. While that implies that I have to do that for every VO, it's much cleaner (basically, what I was proposing, just not automagically).

Thanks!.

**Comment by jeffward01** [comment]
I have to test this and see if it works for me, but I saw in @Blackclaws  examples, that you did `public partial class VogenValueObjectName { } `

Will it also work for:

`public readonly partial struct VogenValueObjectName { } `

I think my issue was that I am using Guid VogenValueObjects as `struct`, and `struct` is initialized to `default`, and not `null`.  And the `default`, will throw an error.

Perhaps I am mistaken, I will have to test this and see.

**Comment by SteveDunn** [comment]
As mentioned, Vogen intentionally makes it difficult to create uninitialized instances. This offers us that offer reassurance in the _Domain_ layer. However, when dealing with _Infrastructure_ layer, it's a PITA!  Personally, I use primitives in the infrastructure and use an explicit _anti corruption layer_ between infrastructure and domain.

The suggestion of using `_isInitialized` will _likely_ work; `_isInitialized` answers the question "has a value been set via the private constructor?" as opposed to "is the current value null or default?". So yes, Reflection can pass null or default via the private constructor, but the answer via `isInitialized` will not vary based on the type of the Value Object itself or the type of the primitive that it wraps.

**Comment by Blackclaws** [comment]
> As mentioned, Vogen intentionally makes it difficult to create uninitialized instances. This offers reassurance in the _Domain_ layer. However, when dealing with _Infrastructure_ layer, it's a PITA! Personally, I use primitives in the infrastructure and use an explicit _anti corruption layer_ between infrastructure and domain where I map primitives to _domain concepts_ (Value Objects).
> 
> The suggestion of using `_isInitialized` will _likely_ work; `_isInitialized` answers the question "has a value been set via the private constructor?" as opposed to "is the current value null or default?". So yes, Reflection can pass null or default via the private constructor, but the answer via `isInitialized` will not vary based on the type of the Value Object itself or the type of the primitive that it wraps.

I actually find Vogen VOs to be a breeze to deal with on the Infrastructure layer. The conversions to EFCore at least have not provided problems so far, and even when reading configuration that has VOs as properties I haven't had issues so far. As long as you explicitly handle the case of "uninitialized" with null I think they work just as well as primitives.

**Comment by Blackclaws** [comment]
One place where it doesn't work well is when you have a configuration class that has a Dictionary<VO, something>. Because the options framework currently can't map nonstandard keys.

**Comment by SteveDunn** [comment]
@Blackclaws - there was an issue raised and fixed recently that might fix this: https://github.com/SteveDunn/Vogen/issues/386

**Comment by SteveDunn** [comment]
> @Blackclaws - there was an issue raised and fixed recently that might fix this: #386

Actually, that's unrelated. But I'll take a look at the options mapping to dictionary keys, thanks for the steer. I wonder if we can get something working with the typeconverters that it generates

**Comment by SteveDunn** [comment]
> > @Blackclaws - there was an issue raised and fixed recently that might fix this: #386
> 
> Actually, that's unrelated. But I'll take a look at the options mapping to dictionary keys, thanks for the steer. I wonder if we can get something working with the typeconverters that it generates

Having taken another look, I don't think it's currently possibly to get binding to create a value object:

https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs#L600

**Comment by SteveDunn** [comment]
> > > @Blackclaws - there was an issue raised and fixed recently that might fix this: #386
> > 
> > 
> > Actually, that's unrelated. But I'll take a look at the options mapping to dictionary keys, thanks for the steer. I wonder if we can get something working with the typeconverters that it generates
> 
> Having taken another look, I don't think it's currently possibly to get binding to create a value object:
> 
> https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs#L600

It's not possible, at least with the current versions of the configuration binding library.

Is it OK to close this now? I don't want to close it if the original question hasn't been sufficiently answered. But it looks like there is a workaround if you absolutely definitely, **must** check to see if a Value Object is initialized before use (which, again, is something that generally isn't needed as Value Objects, in their anticipated usage in Domains, are pretty much guaranteed to be initialized).

**Comment by CheloXL** [comment]
I would say yes... I don't have anything else to add, nor need any more information. Thank you!
